### ISSUE TITLE
Registering nullable value types doesn't allow inferring the Graph Type

### ISSUE THREAD

**Issue by ebekker** [graphql-dotnet/graphql-dotnet#1604]
## Summary
I tried to register with `GraphTypeTypeRegistry` the nullable variations of the built-in value types to the corresponding `*GraphTypes` (essentially duplicating [this built-in code](https://github.com/graphql-dotnet/graphql-dotnet/blob/master/src/GraphQL/Utilities/GraphTypeTypeRegistry.cs) but for the nullable CLR types) in the hopes that it would be able to _infer_ the types of my fields without explicitly specifying the Graph type.

Unfortunately, it only works if the GraphQL field is itself declared nullable.  If the GraphQL field is declared not nullable, but I have a nullable CLR type field that it maps to, I have to explicitly repeat the type mapping in the field declaration.  Is this by design?  Any way to overcome this?

### Environment (if relevant)
* GraphQL 3.0.0-preview-1490
* .NET SDK 3.1.200-preview-014883
* Windows Server 2016

## Examples

Let's say I have an entity class like so:

```csharp
public class Foobar
{
    public int Id { get; set; }
    public string Name { get; set; }
    public bool? Disabled { get; set; }
}
```

And I have a Graph type class that maps to it:

``` csharp
public class FoobarGraph : ObjectGraphType<Foobar>
{
    public FoobarGraph()
    {
        Field(t => t.Id);
        Field(t => t.Name);

        // This generates an exception, unable to infer, unless I also add
        // type: typeof(BooleanGraphType)
        Field(t => t.Disabled,
            nullable: false);
    }
}
```

**Comment by sungam3r** [comment]
>  Is this by design? 

Yes, it is. Another thing is that such a design can hardly be considered good. There is still a backward compatibility issue. Looking through code I see that you can set `nullable: true` to solve it. Using `type: typeof(BooleanGraphType)` is another option to achive the same. Try it. I will leave this issue open as a reminder to change `Field<T>` API.

**Comment by sungam3r** [comment]
Relates to [comment](https://github.com/graphql-dotnet/graphql-dotnet/issues/369#issuecomment-314830399) about API for definining fields.

@joemcbride What are you think about refactor fields API? Optional parameters cause difficulties for many users, since the effect of their use is not obvious. Many overloads complicates the situation. I would try to somehow simplify the API if you do not mind.

**Comment by HrvojeFER** [comment]
I'm having this exact same problem. 

`
            Field<StringGraphType>(
                "id",
                resolve: context => context.Source.Guid?.ToString());
`

The code above is written without field expressions because GraphQL throws an error saying it can't infer the field name from the expression even when I specify the field name later in the FieldBuilder. I think this is ok, because, as far as I understand, that syntax is supposed to be a shorthand for inferring field names from the given expression and it is limited to simple POCO field getter expressions.

The problem I have with this is that if I want to use resolve: context syntax to resolve fields, I can't specify if the field is nullable. It is not a big problem, because I have services that handle it at the backend, but it would be nice if you could specify nullability with this syntax so my GraphQL Schema could be more precise.
Maybe just add a nullable: false argument to Field methods with resolve: context syntax?

**Comment by sungam3r** [comment]
try
`Field<StringGraphType>( "id", resolve: context => context.Source.Guid?.ToString());`
or
`Field<NonNullGraphType<StringGraphType>>( "id", resolve: context => context.Source.Guid?.ToString());`

**Comment by HrvojeFER** [comment]
Thank you. Sorry for being ignorant. Didn't know there was a `NonNullGraphType<>`.

**Comment by sungam3r** [comment]
@ebekker I hope you managed to solve this issue. For now v7 has improved possibilities for inferring types and building schemas, see readme and migration notes. Also we moved to fluent APIs for creating fields and deprecated a bunch of other APIs. Closing.
